presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/ShardPredicate.java

```diff
@@ -108,54 +108,59 @@ public static ShardPredicate create(TupleDomain<RaptorColumnHandle> tupleDomain,
                 continue;
             }
 
+            StringJoiner columnPredicate = new StringJoiner(" OR ", "(", ")").setEmptyValue("true");
             Ranges ranges = domain.getValues().getRanges();
 
-            // TODO: support multiple ranges
-            if (ranges.getRangeCount() != 1) {
+            // prevent generating complicated metadata queries
+            if (ranges.getRangeCount() > MAX_RANGE_COUNT) {
                 continue;
             }
-            Range range = getOnlyElement(ranges.getOrderedRanges());
 
-            Object minValue = null;
-            Object maxValue = null;
-            if (range.isSingleValue()) {
-                minValue = range.getSingleValue();
-                maxValue = range.getSingleValue();
-            }
-            else {
-                if (!range.getLow().isLowerUnbounded()) {
-                    minValue = range.getLow().getValue();
+            for (Range range : ranges.getOrderedRanges()) {
+                StringJoiner rangePredicate = new StringJoiner(" AND ").setEmptyValue("true");
+                Object minValue = null;
+                Object maxValue = null;
+                if (range.isSingleValue()) {
+                    minValue = range.getSingleValue();
+                    maxValue = range.getSingleValue();
                 }
-                if (!range.getHigh().isUpperUnbounded()) {
-                    maxValue = range.getHigh().getValue();
+                else {
+                    if (!range.getLow().isLowerUnbounded()) {
+                        minValue = range.getLow().getValue();
+                    }
+                    if (!range.getHigh().isUpperUnbounded()) {
+                        maxValue = range.getHigh().getValue();
+                    }
                 }
-            }
 
-            String min;
-            String max;
-            if (handle.isBucketNumber()) {
-                if (!bucketed) {
-                    predicate.add("false");
-                    continue;
+                String min;
+                String max;
+                if (handle.isBucketNumber()) {
```

Comment:
This logic is confusing (the `break` change makes it worse). We should fix this by only adding a bucket number column to bucketed tables in `RaptorMetadata`. Then we can remove this `!bucketed` special case.

Do this as a separate commit before this one.